### PR TITLE
Exclude overriden properties from Property Descriptions section

### LINKED ISSUE #76971: ImageTexture class reference has a pointless section
### Godot version

4.0.2

### System information

Ubuntu 22.04.1 LTS

### Issue description

image

Normally, the "Property descriptions" section would be skipped in this case, but because there's an overridden property (which doesn't need a description) this section shows up without needing to.

### Steps to reproduce

Open ImageTexture's class reference.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by brno32**
Fixes https://github.com/godotengine/godot/issues/76971

### CODE REVIEW

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -915,13 +915,11 @@ void EditorHelp::_update_doc() {
...
 
-	bool has_properties = cd.properties.size() != 0;
-	if (cd.is_script_doc) {
-		has_properties = false;
-		for (int i = 0; i < cd.properties.size(); i++) {
-			if (cd.properties[i].name.begins_with("_") && cd.properties[i].description.strip_edges().is_empty()) {
-				continue;
-			}
+	bool has_properties = false;
+	for (int i = 0; i < cd.properties.size(); i++) {
```
Comment (AThousandShips) [inline editor/editor_help.cpp:919]: I'd suggest using foreach syntax here instead:
```for (const DocData::PropertyDoc &prop : cd.properties)```
and then replacing `cd.properties[i]` with `prop`

It's more readable and compact

**Comment by AThousandShips** [comment]
To link it to the issue write is as "Fixes #76971"

File: editor/editor_help.cpp (hunk 1 above)
Comment (brno32) [reply editor/editor_help.cpp]: agreed!

File: editor/editor_help.cpp [hunk 2]
```diff
@@ -915,13 +915,11 @@ void EditorHelp::_update_doc() {
...
-	bool has_properties = cd.properties.size() != 0;
-	if (cd.is_script_doc) {
-		has_properties = false;
-		for (int i = 0; i < cd.properties.size(); i++) {
-			if (cd.properties[i].name.begins_with("_") && cd.properties[i].description.strip_edges().is_empty()) {
-				continue;
-			}
+	bool has_properties = false;
+	for (const DocData::PropertyDoc &prop : cd.properties) {
+		if (cd.is_script_doc && prop.name.begins_with("_") && prop.description.strip_edges().is_empty()) {
```
Comment (AThousandShips) [inline editor/editor_help.cpp:920]: I'd suggest combining this into `prop.overridden || ...` and then just removing the second if, don't think scripts can override properties anyway (the override here just deals with the default value as far as I know)

**Comment by AThousandShips** [comment]
Note that I'm not able to review this properly, you need someone from the editor team who have write access

File: editor/editor_help.cpp (hunk 2 above)
Comment (brno32) [reply editor/editor_help.cpp:920]: I almost want to move `prop.name.begins_with("_") && prop.description.strip_edges().is_empty()` to some method on the `PropertyDoc` class to make this more readable. Like

`cd.is_script_doct && prop.is_undocumented_private` or something like that

Comment (AThousandShips) [reply editor/editor_help.cpp]: Is it used anywhere else? If not it's not really worth it I'd say

Comment (brno32) [reply editor/editor_help.cpp]: doesn't look like it. I updated the code!

Comment (AThousandShips) [reply editor/editor_help.cpp]: The grouping is not needed, && has higher precedence

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -915,16 +915,12 @@ void EditorHelp::_update_doc() {
...
-				continue;
-			}
-			has_properties = true;
-			break;
+	bool has_properties = false;
+	for (const DocData::PropertyDoc &prop : cd.properties) {
+		if (prop.overridden || (cd.is_script_doc && prop.name.begins_with("_") && prop.description.strip_edges().is_empty())) {
+			continue;
 		}
+		has_properties = true;
```
Comment (AThousandShips) [inline editor/editor_help.cpp:923]: You should break here, no reason to continue iteration

**Comment by AThousandShips** [comment]
As I said I can't help you reviewing this beyond giving suggestions, someone will come along and assign the correct team to this who can approve this, but I get pinged each time you re-request a review

**Comment by brno32** [comment]
> As I said I can't help you reviewing this beyond giving suggestions, someone will come along and assign the correct team to this who can approve this, but I get pinged each time you re-request a review

Sorry. Hitting that button is a habit 😄 Thanks!

**Comment by kleonc** [comment]
> As I said I can't help you reviewing this beyond giving suggestions, someone will come along and assign the correct team to this who can approve this, but I get pinged each time you re-request a review

@AThousandShips To be clear, you can review/approve other people's PRs like any other user (not blocked by [godotengine](https://github.com/godotengine))(see: [Approving a pull request with required reviews](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews)). The difference is: an approval by non-member (of [godotengine](https://github.com/godotengine)) doesn't allow merging the PR, as at least one member-approval is required (with exception for the owners I think (you know, the super-members:smile:), they could merge anything as they like (I mean technically they have such power)). But a non-member's review (approval is a form of review) is still a valuable feedback (especially when they gave some suggestions; like after seeing an approval following some changes I'd assume their previous suggestions/comments were addressed).

**Comment by AThousandShips** [comment]
@kleonc thank you, I do understand this and thank you for explaining how it does add that depth to a review, will keep this in mind, the reason I haven't approved any changes is that while I have knowledge on some of the codebase and c++ I don't feel confident saying this is the way to go etc.

**Review by AThousandShips — APPROVED** [review]

File: editor/editor_help.cpp [hunk 4]
```diff
@@ -915,16 +915,13 @@ void EditorHelp::_update_doc() {
...
+		break;
```
Comment (kleonc) [inline editor/editor_help.cpp:924]: I think this:
```suggestion
	bool has_properties = false;
	bool has_property_descriptions = false;
	for (const DocData::PropertyDoc &prop : cd.properties) {
		if (cd.is_script_doc && prop.name.begins_with("_") && prop.description.strip_edges().is_empty()) {
			continue;
		}
		has_properties = true;
		if (!prop.overridden) {
			has_property_descriptions = true;
			break;
		}
```
+changing:
https://github.com/godotengine/godot/blob/fb10f45efe52e330496e325b6b7efd0e52f769b0/editor/editor_help.cpp#L1603-L1604
to check `has_property_descriptions` should do the thing.

**Review by kleonc — CHANGES_REQUESTED** [review]
Currently `has_properties` is used as a condition for whether to add both `Properties` and `Property Descriptions` sections. Meaning currently this PR makes both these sections gone in case of only overridden properties. Both sections need separate conditions for showing them.

**Comment by kleonc** [comment]
> @kleonc thank you, I do understand this and thank you for explaining how it does add that depth to a review, will keep this in mind, the reason I haven't approved any changes is that while I have knowledge on some of the codebase and c++ I don't feel confident saying this is the way to go etc.

To clarify further: personally I usually try to not approve if I'm not sure about the correctness. Or point out I'm not sure / my doubts in case I'm approving as "it should be good" etc. By any means I'm not saying you need to approve after you've commented on something. But if you think so, go ahead, that's also a feedback. :slightly_smiling_face:

**Comment by AThousandShips** [comment]
@kleonc thank you, will add that in the future!

**Comment by AThousandShips** [comment]
You didn't update the use of `has_properties` as per the suggestion:
> +changing:
https://github.com/godotengine/godot/blob/fb10f45efe52e330496e325b6b7efd0e52f769b0/editor/editor_help.cpp#L1603-L1604
> 
> to check has_property_descriptions should do the thing.

**Comment by brno32** [comment]
> You didn't update the use of `has_properties` as per the suggestion:
> 
> > +changing:
> > https://github.com/godotengine/godot/blob/fb10f45efe52e330496e325b6b7efd0e52f769b0/editor/editor_help.cpp#L1603-L1604
> > 
> > to check has_property_descriptions should do the thing.

You're too quick for me :D I committed from the GitHub UI but haven't pulled down the changes and updated the code yet

**Comment by AThousandShips** [comment]
My bad! I assumed you'd add it manually if you were adding the additional code

**Comment by brno32** [comment]
> My bad! I assumed you'd add it manually if you were adding the additional code, especially since it runs CI needlessly

Fair enough. That's something to consider

> Currently `has_properties` is used as a condition for whether to add both `Properties` and `Property Descriptions` sections. Meaning currently this PR makes both these sections gone in case of only overridden properties. Both sections need separate conditions for showing them.

Thank you for catching this. I see I misunderstood the original intention. We only want to show the Property Description section if there are non-overriden properties in the class.

**Comment by AThousandShips** [comment]
In my original suggestion for a fix for this in the issue I wasn't sure exactly what was going on and had missed that it was used to also control the list of properties separate from descriptions, didn't investigate how it worked so wasn't catching that part

**Comment by kleonc** [comment]
Well, I was looking at this code for the first time myself, so it's not like I'm familiar with it. :upside_down_face: Just checked where `has_properties` was being used. And looking at [the image](https://github.com/godotengine/godot/assets/85438892/47baf956-c28c-4f8f-aed5-c515528d4ed6) from #76971 the `Properties` section looks fine to me.

**Review by kleonc — APPROVED** [review]
LGTM:
|Before|After|
|-|-|
|ZdTlKpewL7|hwxjpJWKjW

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.3.

**Pull request merged**